### ISSUE TITLE
Basic auth failed

### ISSUE THREAD

**Issue by pavel-jurasek-merck-com** [modauthgssapi/mod_auth_gssapi#28]
@simo5 i figure out delegation credentials now i am stuck with fallback basic auth not working :/ Still getting 401 even if i use correct username and password:/

Apache2 config
```
<VirtualHost *:80>
	ServerName  kerberos.mywork.merck.com
        ServerAlias 54.40.221.55

	ServerAdmin pavel.jurasek@merck.com
	DocumentRoot /var/www/html

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

 RewriteEngine On

 RewriteCond %{DOCUMENT_ROOT}/%{REQUEST_FILENAME} !-f
 RewriteRule ^/(.*)$ balancer://upstream%{REQUEST_URI} [P,QSA,L]

     <Location />
    AuthType GSSAPI
    AuthName "Login"
    Require valid-user

    GssapiUseS4U2Proxy On
    GssapiCredStore keytab:/etc/apache2/kerberos.mywork.service.keytab
    GssapiCredStore client_keytab:/etc/apache2/kerberos.mywork.service.keytab
    GssapiDelegCcacheDir /home/admin
    GssapiBasicAuth On
    RequestHeader set REMOTE_USER %{REMOTE_USER}s
    RequestHeader set AUTH_TYPE %{AUTH_TYPE}s
    RequestHeader set KRB5CCNAME %{KRB5CCNAME}s
</Location>

<Proxy balancer://upstream>
  BalancerMember http://0.0.0.0:9999
 </Proxy>

 ProxyRequests Off
 ProxyVia On
 ProxyPreserveHost On

 <Proxy *>
  Require all granted
 </Proxy>

 <Directory /var/www/html>
  Options -MultiViews
  Require all granted
 </Directory>

</VirtualHost>
```

Local klist -f -> verify that i do not have ticket

```
klist -f
Credentials cache: API:1C2F03F8-3ECC-4EE4-B82E-F616A881F691
        Principal: jurasek@MERCK.COM

  Issued    Expires  Flags    Principal
```

Curl to server:

```
curl --verbose -u <username>:<password> http://kerberos.mywork.merck.com
* Adding handle: conn: 0x7fec40804000
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x7fec40804000) send_pipe: 1, recv_pipe: 0
* About to connect() to kerberos.mywork.merck.com port 80 (#0)
*   Trying 54.40.222.179...
* Connected to kerberos.mywork.merck.com (54.40.222.179) port 80 (#0)
* Server auth using Basic with user '<username>'
> GET / HTTP/1.1
> Authorization: Basic PHVzZXJuYW1lPjo8cGFzc3dvcmQ+
> User-Agent: curl/7.30.0
> Host: kerberos.mywork.merck.com
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< Content-Type: text/html; charset=iso-8859-1
< Date: Mon, 11 May 2015 20:01:13 GMT
* Server Apache/2.4.10 (Debian) is not blacklisted
< Server: Apache/2.4.10 (Debian)
< WWW-Authenticate: Negotiate
* Authentication problem. Ignoring this.
< WWW-Authenticate: Basic realm="Login"
< Content-Length: 472
< Connection: keep-alive
<
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>401 Unauthorized</title>
</head><body>
<h1>Unauthorized</h1>
<p>This server could not verify that you
are authorized to access the document
requested.  Either you supplied the wrong
credentials (e.g., bad password), or your
browser doesn't understand how to supply
the credentials required.</p>
<hr>
<address>Apache/2.4.10 (Debian) Server at kerberos.mywork.merck.com Port 80</address>
</body></html>
* Connection #0 to host kerberos.mywork.merck.com left intact
```

Apache2 - Access log

```
54.40.222.179 - - [11/May/2015:20:01:13 +0000] "GET / HTTP/1.1" 401 760 "-" "curl/7.30.0"
```

Apache2 - Error log

```
[Mon May 11 20:01:13.248338 2015] [auth_gssapi:error] [pid 13516] [client 54.40.222.179:3540] gss_init_sec_context() failed: [Unspecified GSS failure.  Minor code may provide more information (Server not found in Kerberos database)]
```

**Comment by rcritten** [comment]
Server not found in Kerberos database. This suggests either your keytab is incorrect or something is wrong in /etc/krb5.conf. The KDC can't find a service for HTTP/kerberos.mywork.merck.com.

**Comment by pavel-jurasek-merck-com** [comment]
@rcritten  @abbra  @simo5 my principal is HTTP/kerberos.mywork.merck.com@MERCK.COM so should be fine then no? I tested with klist :

```
Ticket cache: FILE:/home/admin/jurasek@MERCK.COM
Default principal: jurasek@MERCK.COM

Valid starting       Expires              Service principal
05/11/2015 20:24:33  05/12/2015 06:24:16  HTTP/kerberos.mywork.merck.com@MERCK.COM
05/11/2015 18:13:47  05/12/2015 04:13:47  krbtgt/MERCK.COM@MERCK.COM
	for client HTTP/kerberos.mywork.merck.com@MERCK.COM, renew until 05/18/2015 18:13:47
```

**Comment by pavel-jurasek-merck-com** [comment]
i using as user who want to join site this name : jurasek without realm i also try jurasek@MERCK.COM ... BTW i am not able to see non error debug lines in code...

My error log looks like this:

[Tue May 12 06:49:50.479668 2015] [rewrite:trace1] [pid 14844] mod_rewrite.c(475): [client 54.40.222.179:15442] 54.40.222.179 - - [kerberos.mywork.merck.com/sid#7fcd55759ad0][rid#7fcd5572a0a0/initial] go-ahead with proxy request proxy:balancer://upstream/ [OK]
[Tue May 12 06:49:50.479725 2015] [authz_core:debug] [pid 14844] mod_authz_core.c(809): [client 54.40.222.179:15442] AH01626: authorization result of Require valid-user : denied (no authenticated user yet)
[Tue May 12 06:49:50.479731 2015] [authz_core:debug] [pid 14844] mod_authz_core.c(809): [client 54.40.222.179:15442] AH01626: authorization result of <RequireAny>: denied (no authenticated user yet)

**Comment by simo5** [comment]
> Apache2 - Error log
> 
> ```
> [Mon May 11 20:01:13.248338 2015] [auth_gssapi:error] [pid 13516] [client
> 54.40.222.179:3540] gss_init_sec_context() failed: [Unspecified GSS failure.
> Minor code may provide more information (Server not found in Kerberos
> database)]
> ```

This error means mod_auth_gssapi tried to "validate" the ticket obtained with
your plain text credentials but could not obtain a ticket to "itself".

The way validation works is that a krbtgt is obtained on the server and then
it is used to request a ticket for the kerberos principal in the provided
keytab. If the keytab is missing or "bad" then this operation will fail.

In this case it seem that either the keytab contains a bad principal name or
you do not have a keytab at all.

-- 
Simo Sorce * Red Hat, Inc. * New York

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 i have keytab on my server eveything work except basic auth fallback.

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 here is my GSSAPI conf again:

    AuthType GSSAPI
    AuthName "MERCK.COM"
    Require valid-user

    GssapiUseS4U2Proxy On
    GssapiCredStore keytab:/etc/apache2/kerberos.mywork.service.keytab
    GssapiCredStore client_keytab:/etc/apache2/kerberos.mywork.service.keytab
    GssapiDelegCcacheDir /home/admin
    GssapiBasicAuth On
    RequestHeader set REMOTE_USER %{REMOTE_USER}s
    RequestHeader set AUTH_TYPE %{AUTH_TYPE}s
    RequestHeader set KRB5CCNAME %{KRB5CCNAME}s
    GssapiConnectionBound On
    Header set Persistent-Auth "true"

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 @abbra @rcritten kinit on server works like charm i getting tickets so that will be not a problem ?

**Comment by simo5** [comment]
2 things you can try:
1. remove the GssapiCredStore client_keytab directive for now
2. on the server:
kinit as your user
then run: kvno HTTP/kerberos.mywork.merck.com@MERCK.COM
then check with klist if you have a ticket for the service and what is the vno
then klist -kt /etc/apache2/kerberos.mywork.service.keytab and check that the vno matches

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 i did what you ask me to do here is output... i commented out client_keytab no luck :/
![screen shot 2015-05-12 at 12 41 34](https://cloud.githubusercontent.com/assets/7910967/7585466/517c89fc-f8a4-11e4-96d5-7c05cff768f8.png)

**Comment by frenche** [comment]
Hi,

Could you try to comment out "GssapiUseS4U2Proxy On" for testing?

Thanks,
Isaac B.

**Comment by frenche** [comment]
@simo5 I think I can reproduce this when 'GssapiUseS4U2Proxy' is on.

It looks like the reason is because when 'use_s4u2proxy' is true we set 'cred_usage' to 'GSS_C_BOTH', later we do 'gss_acquire_cred[from]' and then we do 'gss_inquire_cred' in order to get the 'server' name which we later use as the target name in 'gss_init_sec_context' call.

But when 'cred_usage' is set to 'GSS_C_BOTH' in my lab I get the *client* name inside the 'server' when calling 'gss_inquire_cred' which then cause  'gss_init_sec_context' to fail with "Server not found in Kerberos database" error.

HTH,
Isaac B.

**Comment by frenche** [comment]
Perhaps related to:
http://krbdev.mit.edu/rt/Ticket/History.html?id=8152

Cheers.

**Comment by simo5** [comment]
It shouldn't as we create a new ccache for "with_password" operation exactly to avoid a host of issues we have with that API.

**Comment by simo5** [comment]
The GSS-C-BOTH issue seem a real problem though, I am thinking of a way to handle it by allowing to specifiy explicitly a SPN in the configuration.

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 do you know some ETA for this feature? :)

**Comment by simo5** [comment]
pavel, can you check this commit and let me know if it fixes the issue ?
https://github.com/simo5/mod_auth_gssapi/commit/66ea56c49e70cbd2caeea1f85cbee204293f9890

**Comment by frenche** [comment]
Hi,

On Mon, May 25, 2015 at 4:58 PM, simo5 <notifications@github.com> wrote:

> pavel, can you check this commit and let me know if it fixes the issue ?
> simo5@66ea56c
> <https://github.com/simo5/mod_auth_gssapi/commit/66ea56c49e70cbd2caeea1f85cbee204293f9890>
>
If I may, maybe it would be simpler to start with
'gss_acquire_cred_from(GSS_C_ACCEPT)' then do the 'inquire' to get the
server name and then add (again inside HAVE_GSS_ACQUIRE_CRED_FROM) 'if
(use_s4u2proxy) ss_add_cred_from(GSS_C_INITIATE)'.
I havn't tested it but it looks somewhat cleaner (to me :).

Regards,
Isaac B.

**Comment by simo5** [comment]
On Mon, 2015-05-25 at 10:07 -0700, Isaac Boukris wrote:

> If I may, maybe it would be simpler to start with
> 'gss_acquire_cred_from(GSS_C_ACCEPT)' then do the 'inquire' to get the
> server name and then add (again inside HAVE_GSS_ACQUIRE_CRED_FROM) 'if
> (use_s4u2proxy) ss_add_cred_from(GSS_C_INITIATE)'.
> I havn't tested it but it looks somewhat cleaner (to me :).

It wouldn't work, you have to acquire the cred with BOTH, as it is a
single credential marked as good for both initiation and accepting, it
is not a set of 2 separate credentials.

-- 
Simo Sorce * Red Hat, Inc * New York

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 i will today around 23:00 CET let you know how it works :)

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 i am not able to compile your version of gssapi :(

This is output once i try to do make:

/usr/lib64/apr-1/build/libtool --silent  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.    -pthread    -I/usr/local/apache2/include  -I/usr/include/apr-1  -g -O2 -MT mod_auth_gssapi_la-mod_auth_gssapi.lo -MD -MP -MF .deps/mod_auth_gssapi_la-mod_auth_gssapi.Tpo -c -o mod_auth_gssapi_la-mod_auth_gssapi.lo `test -f 'mod_auth_gssapi.c' || echo './'`mod_auth_gssapi.c
In file included from mod_auth_gssapi.c:25:0:
mod_auth_gssapi.h:44:5: error: unknown type name ‘gss_key_value_set_desc’
     gss_key_value_set_desc *cred_store;
     ^
mod_auth_gssapi.c: In function ‘mag_store_deleg_creds’:
mod_auth_gssapi.c:125:5: error: unknown type name ‘gss_key_value_element_desc’
     gss_key_value_element_desc element;
     ^
mod_auth_gssapi.c:126:5: error: unknown type name ‘gss_key_value_set_desc’
     gss_key_value_set_desc store;
     ^
mod_auth_gssapi.c:137:12: error: request for member ‘key’ in something not a structure or union
     element.key = "ccache";
            ^
mod_auth_gssapi.c:138:12: error: request for member ‘value’ in something not a structure or union
     element.value = value;
            ^
mod_auth_gssapi.c:139:10: error: request for member ‘elements’ in something not a structure or union
     store.elements = &element;
          ^
mod_auth_gssapi.c:140:10: error: request for member ‘count’ in something not a structure or union
     store.count = 1;
          ^
mod_auth_gssapi.c: In function ‘mag_cred_store’:
mod_auth_gssapi.c:469:5: error: unknown type name ‘gss_key_value_element_desc’
     gss_key_value_element_desc *elements;
     ^
In file included from /usr/include/apr-1/apr_strings.h:50:0,
                 from mod_auth_gssapi.h:11,
                 from mod_auth_gssapi.c:25:
mod_auth_gssapi.c:493:46: error: ‘gss_key_value_set_desc’ undeclared (first use in this function)
                                       sizeof(gss_key_value_set_desc));
                                              ^
/usr/include/apr-1/apr_pools.h:465:51: note: in definition of macro ‘apr_pcalloc’
 #define apr_pcalloc(p, size) memset(apr_palloc(p, size), 0, size)
                                                   ^
mod_auth_gssapi.c:493:46: note: each undeclared identifier is reported only once for each function it appears in
                                       sizeof(gss_key_value_set_desc));
                                              ^
/usr/include/apr-1/apr_pools.h:465:51: note: in definition of macro ‘apr_pcalloc’
 #define apr_pcalloc(p, size) memset(apr_palloc(p, size), 0, size)
                                                   ^
mod_auth_gssapi.c:499:23: error: ‘gss_key_value_element_desc’ undeclared (first use in this function)
         size = sizeof(gss_key_value_element_desc) * MAX_CRED_OPTIONS;
                       ^
mod_auth_gssapi.c:500:24: error: request for member ‘elements’ in something not a structure or union
         cfg->cred_store->elements = apr_palloc(parms->pool, size);
                        ^
mod_auth_gssapi.c:501:29: error: request for member ‘elements’ in something not a structure or union
         if (!cfg->cred_store->elements) {
                             ^
mod_auth_gssapi.c:507:31: error: request for member ‘elements’ in something not a structure or union
     elements = cfg->cred_store->elements;
                               ^
mod_auth_gssapi.c:508:28: error: request for member ‘count’ in something not a structure or union
     count = cfg->cred_store->count;
                            ^
mod_auth_gssapi.c:516:20: error: request for member ‘count’ in something not a structure or union
     cfg->cred_store->count++;
                    ^
mod_auth_gssapi.c:518:20: error: request for member ‘key’ in something not a structure or union
     elements[count].key = key;
                    ^
mod_auth_gssapi.c:519:20: error: request for member ‘value’ in something not a structure or union
     elements[count].value = value;
                    ^
make[2]: *** [mod_auth_gssapi_la-mod_auth_gssapi.lo] Error 1
make[2]: Leaving directory `/home/ec2-user/testik/mod_auth_gssapi/src'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/ec2-user/testik/mod_auth_gssapi/src'
make: *** [all-recursive] Error 1
[ec2-user@ip-54-40-221-178 mod_auth_gssapi]$

Version from this repo works without problem.

**Comment by pavel-jurasek-merck-com** [comment]
[ec2-user@ip-54-40-221-178 mod_auth_gssapi]$ /usr/local/apache2/bin/httpd -V
Server version: Apache/2.4.12 (Unix)
Server built:   May  6 2015 10:27:39
Server's Module Magic Number: 20120211:41
Server loaded:  APR 1.5.0, APR-UTIL 1.4.1
Compiled using: APR 1.5.0, APR-UTIL 1.4.1
Architecture:   64-bit
Server MPM:     event
  threaded:     yes (fixed thread count)
    forked:     yes (variable process count)
Server compiled with....
 -D APR_HAS_SENDFILE
 -D APR_HAS_MMAP
 -D APR_HAVE_IPV6 (IPv4-mapped addresses enabled)
 -D APR_USE_SYSVSEM_SERIALIZE
 -D APR_USE_PTHREAD_SERIALIZE
 -D SINGLE_LISTEN_UNSERIALIZED_ACCEPT
 -D APR_HAS_OTHER_CHILD
 -D AP_HAVE_RELIABLE_PIPED_LOGS
 -D DYNAMIC_MODULE_LIMIT=256
 -D HTTPD_ROOT="/usr/local/apache2"
 -D SUEXEC_BIN="/usr/local/apache2/bin/suexec"
 -D DEFAULT_PIDLOG="logs/httpd.pid"
 -D DEFAULT_SCOREBOARD="logs/apache_runtime_status"
 -D DEFAULT_ERRORLOG="logs/error_log"
 -D AP_TYPES_CONFIG_FILE="conf/mime.types"
 -D SERVER_CONFIG_FILE="conf/httpd.conf"

**Comment by simo5** [comment]
pavel-jurasek-merck-com it looks like you have an older version of the Kerberos libraries, or are missing the development headers.
You need MIT Kerberos 1.11 at least, but I think you want at least 1.12 to test the s4u2proxy part.
You may also want to install gss-ntlmssp and headers too if you are trying my latest branches as I am adding a build dependency (not runtime) to it.

**Comment by pavel-jurasek-merck-com** [comment]
@simo5 it's working now! Thank you:]

curl -u jurasek:Jj123456 --verbose http://kerberos.mywork.merck.com/
* Adding handle: conn: 0x7fd522004000
* Adding handle: send: 0
* Adding handle: recv: 0
* Curl_addHandleToPipeline: length: 1
* - Conn 0 (0x7fd522004000) send_pipe: 1, recv_pipe: 0
* About to connect() to proxy beantc.merck.com port 8081 (#0)
*   Trying 54.101.71.2...
* Connected to beantc.merck.com (54.101.71.2) port 8081 (#0)
* Server auth using Basic with user 'jurasek'
> GET http://kerberos.mywork.merck.com/ HTTP/1.1
> Authorization: Basic <pass>
> User-Agent: curl/7.30.0
> Host: kerberos.mywork.merck.com
> Accept: */*
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 OK
< Via: 1.1 kerberos.mywork.merck.com
< Date: Tue, 26 May 2015 12:02:15 GMT
< Vary: Accept-Encoding
* Server Apache/2.4.10 (Debian) is not blacklisted
< Server: Apache/2.4.10 (Debian)
< Status: 200 OK
< X-Cache: MISS from 54.101.240.111
< Set-Cookie: LBN=node1; path=/
< Content-Type: text/html;charset=utf-8
< Cache-control: private
< Content-Length: 962
< X-Frame-Options: SAMEORIGIN
< Proxy-Connection: Keep-Alive
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
<
* Connection #0 to host beantc.merck.com left intact

<content of page>

**Comment by simo5** [comment]
Thank you for testing

**Comment by frenche** [comment]
Hi,

On Mon, May 25, 2015 at 4:34 PM, simo5 <notifications@github.com> wrote:

> The GSS-C-BOTH issue seem a real problem though, I am thinking of a way to
> handle it by allowing to specifiy explicitly a SPN in the configuration.
>

Hmm, this brings another idea, maybe we could acquire credentials and
inquire them once at 'mag_use_basic_auth()' and then save the server
name SPN  so we don't need to acquire (sometimes twice) at runtime.
The only difference I see is that changing the keytab content will require
reconf of the server (right?).
Not sure how costly it currently is (if at all) but if you deem appropriate
I can try to make an alternative patch.

Regards,
Isaac B.

**Comment by simo5** [comment]
For basic auth it would probably be fine, as long as we do not leak it to ther auth types.
(Please open a new issue to discuss this if you want to bring this idea forward.)
